pkg/chains/formats/slsa/v2alpha2/internal/resolved_dependencies/resolved_dependencies.go

```diff
@@ -0,0 +1,331 @@
+	var resolvedDependencies []v1.ResourceDescriptor
+
+	// add top level task config
+	if p := tro.Status.Provenance; p != nil && p.RefSource != nil {
+		rd := v1.ResourceDescriptor{
+			Name:   "config/task",
+			URI:    p.RefSource.URI,
+			Digest: p.RefSource.Digest,
+		}
+		resolvedDependencies = append(resolvedDependencies, rd)
+	}
+
+	// add step images
+	if err := addStepImagesToResolvedDependencies(tro.Status.Steps, &resolvedDependencies); err != nil {
+		return resolvedDependencies, nil
+	}
+
+	// add sidecar images
+	if err := addSidecarImagesToResolvedDependencies(tro.Status.Sidecars, &resolvedDependencies); err != nil {
+		return resolvedDependencies, nil
+	}
+
+	gitCommit, gitURL := gitInfo(tro)
+
+	// Store git rev as ResolvedDependencies
+	if gitCommit != "" && gitURL != "" {
+		resolvedDependencies = append(resolvedDependencies, v1.ResourceDescriptor{
+			Name:   "inputs/result",
+			URI:    gitURL,
+			Digest: map[string]string{"sha1": gitCommit},
+		})
+		return resolvedDependencies, nil
+	}
+
+	// add input artifacts
+	sms := artifacts.RetrieveResolvedDependenciesFromStructuredResults(ctx, tro, artifacts.ArtifactsInputsResultName)
+	for i := range sms {
+		sms[i].Name = "inputs/result"
+	}
+	resolvedDependencies = append(resolvedDependencies, sms...)
+
+	if tro.Spec.Resources != nil { //nolint:all //incompatible with pipelines v0.45
+		// check for a Git PipelineResource
+		for _, input := range tro.Spec.Resources.Inputs { //nolint:all //incompatible with pipelines v0.45
+			if input.ResourceSpec == nil || input.ResourceSpec.Type != backport.PipelineResourceTypeGit { //nolint:all //incompatible with pipelines v0.45
+				continue
+			}
+
+			rd := v1.ResourceDescriptor{
+				Name:   "pipelineResource",
+				Digest: common.DigestSet{},
+			}
+
+			for _, rr := range tro.Status.ResourcesResult {
+				if rr.ResourceName != input.Name {
+					continue
+				}
+				if rr.Key == "url" {
+					rd.URI = attest.SPDXGit(rr.Value, "")
+				} else if rr.Key == "commit" {
+					rd.Digest["sha1"] = rr.Value
+				}
+			}
+
+			var url string
+			var revision string
+			for _, param := range input.ResourceSpec.Params {
+				if param.Name == "url" {
+					url = param.Value
+				}
+				if param.Name == "revision" {
+					revision = param.Value
+				}
+			}
+			rd.URI = attest.SPDXGit(url, revision)
+			resolvedDependencies = append(resolvedDependencies, rd)
+		}
+	}
+
+	// remove duplicate resolved dependencies
+	resolvedDependencies, err := removeDuplicateResolvedDependencies(resolvedDependencies)
+	if err != nil {
+		return resolvedDependencies, err
+	}
+	return resolvedDependencies, nil
+}
+
+// PipelineRunResolvedDependencies constructs `predicate.resolvedDependencies` section by collecting all the artifacts that influence a taskrun such as source code repo and step&sidecar base images.
+func PipelineRunResolvedDependencies(ctx context.Context, pro *objects.PipelineRunObject) ([]v1.ResourceDescriptor, error) {
```

Comment:
I have 2 comments:
- these 2 functions `PipelineRunResolvedDependencies` and `TaskRunResolvedDependencies` are too long, which makes it hard to read and understand what are put into resolved dependencies. Can we split them into smaller functions and call them in `PipelineRunResolvedDependencies`? i.e. extract from tasks for base images, extract from pipeline level results?
- It might be helpful for readers to understand the file quickly by putting those 2 important functions at the top of the file.

Reply:
done

Reply:
Can we also move the 2 functions to the top of the file as well? Thanks!

Reply:
sure!

Reply:
done!